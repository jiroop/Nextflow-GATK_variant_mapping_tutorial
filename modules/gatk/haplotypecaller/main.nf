// Call variants with GATK HaplotypeCaller

process GATK_HAPLOTYPECALLER {

    container 'quay.io/biocontainers/gatk4:4.6.2.0--py310hdfd78af_1'

    publishDir params.outdir, mode: 'symlink'

    input:
        tuple path (input_bam), path(input_bam_index)
        path ref_fasta
        path ref_index
        path ref_dict
        path interval_list

    output:
        path "${input_bam}.g.vcf"     , emit: vcf
        path "${input_bam}.g.vcf.idx" , emit: idx // this is the index file for the g.vcf that is automatically created by GATK

    script:
    """
    gatk HaplotypeCaller \
    -R ${ref_fasta} \
    -I ${input_bam} \
    -O ${input_bam}.g.vcf \
    -L ${interval_list} \
    -ERC GVCF 

    #.g.vcf for genomic vcf
    # The -ERC GVCF switches the haplotype caller to GVCF mode for doing genomic VCF calling for joint genotyping analysis 

    """
}