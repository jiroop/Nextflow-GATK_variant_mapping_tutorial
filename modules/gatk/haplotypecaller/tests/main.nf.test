nextflow_process {

    name "Test Process GATK_HAPLOTYPECALLER"
    script "modules/gatk/haplotypecaller/main.nf"
    process "GATK_HAPLOTYPECALLER"

    test("Should call son's haplotype correctly") {

        setup {
            run("SAMTOOLS_INDEX") {
                script "../../../samtools/index/main.nf"
                process {
                    """
                    input[0] =  file("${projectDir}/data/bam/reads_son.bam")
                    """
                }
            }
        }

        when {
            params {
                outdir = "tests/results"
            }
            process {
                """
                input[0] = SAMTOOLS_INDEX.out
                input[1] = file("${projectDir}/data/ref/ref.fasta")
                input[2] = file("${projectDir}/data/ref/ref.fasta.fai")
                input[3] = file("${projectDir}/data/ref/ref.dict")
                input[4] = file("${projectDir}/data/ref/intervals.bed")
                """
            }
        }

        then {
            assert process.success
            // the next two lines test the identify of specific expected lines in the output g.vcf file
            assert path(process.out[0][0]).readLines().contains('#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	reads_son')
            assert path(process.out[0][0]).readLines().contains('20_10037292_10066351	3277	.	G	<NON_REF>	.	.	END=3282	GT:DP:GQ:MIN_DP:PL	0/0:25:72:24:0,72,719')
            // assert snapshot(process.out).match() // this will not working with haplotype caller as it includes timestamps in the output VCF and this will never match the prior snapshot

        }

    }


    test("Should call mothers haplotype correctly") {

        setup {
            run("SAMTOOLS_INDEX") {
                script "../../../samtools/index/main.nf"
                process {
                    """
                    input[0] =  file("${projectDir}/data/bam/reads_mother.bam")
                    """
                }
            }
        }

        when {
            params {
                outdir = "tests/results"
            }
            process {
                """
                input[0] = SAMTOOLS_INDEX.out
                input[1] = file("${projectDir}/data/ref/ref.fasta")
                input[2] = file("${projectDir}/data/ref/ref.fasta.fai")
                input[3] = file("${projectDir}/data/ref/ref.dict")
                input[4] = file("${projectDir}/data/ref/intervals.bed")
                """
            }
        }

        then {
            assert process.success
            // the next two lines test the identify of specific expected lines in the output g.vcf file
            assert path(process.out[0][0]).readLines().contains('#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	reads_mother')
            assert path(process.out[0][0]).readLines().contains('20_10037292_10066351	3277	.	G	<NON_REF>	.	.	END=3278	GT:DP:GQ:MIN_DP:PL	0/0:38:99:37:0,102,1530')

        }

    }

    test("Should call fathers haplotype correctly") {

        setup {
            run("SAMTOOLS_INDEX") {
                script "../../../samtools/index/main.nf"
                process {
                    """
                    input[0] =  file("${projectDir}/data/bam/reads_father.bam")
                    """
                }
            }
        }

        when {
            params {
                outdir = "tests/results"
            }
            process {
                """
                input[0] = SAMTOOLS_INDEX.out
                input[1] = file("${projectDir}/data/ref/ref.fasta")
                input[2] = file("${projectDir}/data/ref/ref.fasta.fai")
                input[3] = file("${projectDir}/data/ref/ref.dict")
                input[4] = file("${projectDir}/data/ref/intervals.bed")
                """
            }
        }

        then {
            assert process.success
            // the next two lines test the identify of specific expected lines in the output g.vcf file
            assert path(process.out[0][0]).readLines().contains('#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	reads_father')
            assert path(process.out[0][0]).readLines().contains('20_10037292_10066351	3277	.	G	<NON_REF>	.	.	END=3281	GT:DP:GQ:MIN_DP:PL	0/0:44:99:42:0,120,1800')

        }

    }

}
