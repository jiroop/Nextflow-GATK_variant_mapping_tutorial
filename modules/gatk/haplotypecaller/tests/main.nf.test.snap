{
    "Should call son's haplotype correctly": {
        "content": [
            {
                "0": [
                    "reads_son.bam.g.vcf:md5,733ac464513466857a82dd5af8ad393e"
                ],
                "1": [
                    "reads_son.bam.g.vcf.idx:md5,e6543fcd9dfa2417fbb044b27de5d3eb"
                ],
                "idx": [
                    "reads_son.bam.g.vcf.idx:md5,e6543fcd9dfa2417fbb044b27de5d3eb"
                ],
                "vcf": [
                    "reads_son.bam.g.vcf:md5,733ac464513466857a82dd5af8ad393e"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.3",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-25T15:27:37.941012"
    }
}